magento/magento2
# Added configuration to enable/disable Clear Shopping Cart action in storefront

### Description (*)

Added admin configuration to enable/disable Clear Shopping Cart action with added confirmation modal when clicked.

The native **Clear Shopping Cart** action is hidden in Luma via CSS, this PR will remove the CSS hidden styling from the theme and create a configuration in the admin panel to control the display.

**Added Configuration:**
- Admin > Stores > Configuration > Sales > Checkout
  - Enable Clear Shopping Cart - (Default value: No)

![enable clear shopping cart config](https://user-images.githubusercontent.com/32643675/83930802-f642e400-a7cb-11ea-90ef-9a0e11aa74b3.png)

![clear shopping cart action](https://user-images.githubusercontent.com/32643675/83931273-68b4c380-a7ce-11ea-8c08-68c7cac980ad.png)

**Added Behaviour:**
- When **Clear Shopping Cart** is clicked, triggers a confirmation widget.


![Clear Shopping cart 1 – 1](https://user-images.githubusercontent.com/50657230/84119805-348f0c00-aa35-11ea-8c81-8c3fd9ad8ab4.png)

- Clicking **OK** will clear your cart and redirect to an empty cart page.

![empty cart page](https://user-images.githubusercontent.com/32643675/83931417-335ca580-a7cf-11ea-9e41-a87b5763e11f.png)

### Fixes Issue(s)
https://github.com/magento/partners-magento2b2b/issues/108

### Related Pull Requests

#27917 

### Manual testing scenarios (*)

1. Go to **Admin** > **Stores** > **Configuration** > **Sales** > **Checkout**
2. Set **Enable Clear Shopping Cart** to **Yes**
3. Click **Save Config**
4. Add a product to cart
5. Visit the Shopping Cart page
6. **Clear Shopping Cart** action should be visible beside the **Update Shopping Cart** action
7. Click **Clear Shopping Cart** and a confirmation modal should popup
8. Click **OK**, you will be redirected to the empty cart page and your cart should now be empty

### Questions or comments

Do these changes require functional tests? If so, I would need assistance with writing the tests.

### Contribution checklist (*)

### Approval Ticket 
https://jira.corp.magento.com/browse/MC-35122

---

### Test Plan
#### Manual checks for ticket
* [x] Verify new configuration option exists in store configuration on Sales > Checkout
  * [x] Check placement of new option (make sure it matches UX mockup)
  * [x] Should be website confg
  * [x] Should be disabled by default w/ "Use system value" checked & default value of No
* [x] Verify can enable/disable and uncheck/check checkbox for new config using UI (saves successfully)
* [x] Verify can enable/disable new config using magento cli
* [x] Verify new config is not available to change at store or storeview scope (change scope in upper left drop-down)
* [x] Saving new config option should trigger configuration cache & page cache to invalidate (& the admin warning message to appear)
* [x] Verify that enabling new config will cause *Clear Shopping Cart* button to appear in cart on storefront
  * [x] Verify that it does **NOT** appear when cart is empty (go to /checkout/cart/index)
* [x] Verify that clicking the button & then OK on modal will empty shopping cart
  * [x] Check for 1 product
  * [x] Check for multiple products w/ multiple quantities
  * [x] Check that mini cart empties as well
  * [x] Edge Case: open cart w/ products, open cart in another browser tab, in tab 2, click clear cart button & accept modal to empty cart, go to tab1 & click clear cart button & accept modal. Verify no errors and user is shown the empty cart page after page reload
* [x] Verify that clicking the button & then clicking cancel on the modal will not change the cart nor reload the page
  * [x] Same for clicking outside modal or hitting Esc on keyboard
* [x] Verify that confirmation modal matches UX mockups
* [x] Verify that disabling new config will cause *Clear Shopping Cart* button to not appear in DOM in cart on storefront
* [x] Verify that deleting the configuration row from core_config_data database table will cause *Clear Shopping Cart* button to not appear in DOM in cart on storefront (simulates upgrade)
* [x] Verify that config can be changed at website level and behaves correctly for websites. Check following scenarios:
  * [x] Disabled for default config, enabled for website: Should be enabled on storefront for website
  * [x] Enabled for default config, disabled for website: Should be disabled on storefront for website
  * [x] Disabled for default config, doesn't exist in db for website: Should be disabled on storefront for website
  * [x] Enabled for default config, doesn't exist in db for website: Should be enabled on storefront for website
  * [x] Doesn't exist in db for for default config, enabled for website: Should be enabled on storefront for website
  * [x] Doesn't exist in db for for default config, doesn't exist in db for website: Should be disabled on storefront for website
* [x] Check translations (of store configuration label, values, checkbox, storefront button text, modal buttons, modal message)
* [x] Check that button on storefront can be focused & clicked using keyboard (accessibility test). Verify modal can be interacted with (cancel & ok) using only keyboard
* [x] Check new button on Luma theme & Blank theme
* [x] Check new button works on Chrome, Safari, Firefox, Edge, iPhone simulator, android simulator
* [x] Verify new config can be changed & saved when Magento is installed w/ database prefix

#### Automated MFTF Tests
* New Test 1
** (see PR MFTF comments)

#### Manual Hip Tests
No new tests
